rllib/core/optim/rl_optimizer.py

```diff
@@ -0,0 +1,94 @@
+import abc
+from typing import Any, List, Mapping
+
+from ray.rllib.core.rl_module import RLModule
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic,
+    OverrideToImplementCustomLogic_CallToSuperRecommended,
+)
+from ray.util.annotations import PublicAPI
+
+
+@PublicAPI(stability="beta")
+class RLOptimizer(abc.ABC):
+    """Base class for defining a loss function and optimizer for a RLModule.
+
+    Args:
+        rl_module: The RLModule that will be optimized.
+        config: The configuration for the optimizer.
+
+    Abstract Methods:
+        compute_loss: computing a loss to optimize rl_module over.
+        _configure_optimizers: constructing the optimizers for rl_module.
+        get_state: getting the state of the optimizer.
+        set_state: setting the state of the optimizer.
+
+    Example:
+    .. code-block:: python
+
+        module = RLModule(...)
+        rl_optim = RLOptimizer(module, config)
+        sample_batch = ...
+        fwd_out = module.forward_train(sample_batch)
+        loss_dict = rl_optim.compute_loss(fwd_out)
+
+        # compute gradients of loss w.r.t. trainable variables
+        ...
+
+        for optim in rl_optim.get_optimizers():
+            optim.step()
+            optim.zero_grad()
+
+    """
+
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def __init__(self, module: RLModule, config: Mapping[str, Any]):
+        self.module = module
+        self._config = config
+        self._optimizers = self._configure_optimizers()
+
+    @abc.abstractmethod
+    def compute_loss(self, fwd_out: Mapping[str, Any]) -> Mapping[str, Any]:
+        """Computes variables for optimizing self._module based on fwd_out.
+
+        Args:
+            fwd_out: Output from a forward pass on self._module during
+                training.
+
+        Returns:
+            A dictionary of tensors used for optimizing self._module.
```

Comment:
The thing that we return here should follow some structure right? It's not just any arbitrary dictionary of tensors. like what is the key that you'd need to be able to optimize? `loss_total`? or do you want to leave it up to the user to define keys here and also in configure_optimizer?

Reply:
I want to leave this up to the user for now

Reply:
Good question. Right now, in torch, we require `optimizer()` to return a list of local optimizers and `loss()` to return the same amount of loss terms.

1) Can we think of any scenarios where we won't have a 1:1 matching between optimizers and losses?

2) If not, I'd be happy with `self._configure_optimizers()` returning a dict (str -> local optimizer) and `self.compute_loss()` to return the same dict struct. We should then validate that these return structs match after the first calls (via `tree.assert_same_structure()`).

Reply:
Also a scenario where you want to return losses just for the sake of logging (not used with any optimizer)

Reply:
ok so instead, lets have 2 returns. 1, our dictionary with extra losses, and 2 our total loss.

Reply:
I think one dict to return is good. Here is my proposal: have the optimizer keys be a subset of keys returned in the loss. 
For the subset keys update the model using losses and their optimizer objects. And log everything including the ones that were not used for optimization. Only error out when there is a key in optimizer object that does not exist in the returned loss dict. If that happened yell at the user why did you not provide the corresponding key. I don't know what I need to do with this optimizer object.

Reply:
The optimizer keys in this design have been agent/policy ids. Sorry for the misunderstanding.

Reply:
I can see that this is the implementation that you have implemented @avnishn. If this was the intention can we clarify this in the documentation of `configure_optimization` and `compute_loss`?

Reply:
ok me and kourosh spoke offline. We'll allow users to return 1 loss tensor or a dictionary containing the protected key "total_loss"